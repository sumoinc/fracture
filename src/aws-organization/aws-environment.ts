import { Component } from "projen";
import { ValueOf } from "type-fest";
import { AwsRegionIdentifier } from "./aws-region";
import { CdkProject } from "../cdk";

export interface IAwsEnvironment {
  readonly orgId: string;
  readonly account: string;
  readonly region: ValueOf<typeof AwsRegionIdentifier>;
}

export interface AwsEnvironmentOptions {
  readonly orgId: string;
  readonly account: string;
  readonly region: ValueOf<typeof AwsRegionIdentifier>;
}

export class AwsEnvironment extends Component {
  public static all(project: CdkProject): Array<AwsEnvironment> {
    const isDefined = (c: Component): c is AwsEnvironment =>
      c instanceof AwsEnvironment;
    return project.components.filter(isDefined);
  }

  /**
   * Make sure this environment is configured for the organization.
   */
  public static ensureExists(
    project: CdkProject,
    options: AwsEnvironmentOptions
  ): AwsEnvironment {
    /**
     * Find function
     */
    const isDefined = (c: Component): c is AwsEnvironment =>
      c instanceof AwsEnvironment &&
      c.orgId === options.orgId &&
      c.account === options.account &&
      c.region === options.region;

    /**
     * It should exist in the project supplied.
     * Return this instance
     */
    return (
      project.components.find(isDefined) ?? new AwsEnvironment(project, options)
    );
  }

  public readonly orgId: string;
  public readonly account: string;
  public readonly region: ValueOf<typeof AwsRegionIdentifier>;

  constructor(
    public readonly project: CdkProject,
    options: AwsEnvironmentOptions
  ) {
    super(project);

    /***************************************************************************
     *
     * DEFAULTS
     *
     **************************************************************************/

    this.orgId = options.orgId;
    this.account = options.account;
    this.region = options.region;
  }

  public config(): IAwsEnvironment {
    return {
      orgId: this.orgId,
      account: this.account,
      region: this.region,
    };
  }
}
