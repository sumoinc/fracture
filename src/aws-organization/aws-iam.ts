/**
 * AWS Managed policies.
 */
export const IamManagedPolicy = {
  ADMINISTRATOR_ACCESS: "AdministratorAccess",
  BILLING: "Billing",
  DATABASE_ADMINISTRATOR: "DatabaseAdministrator",
  DATA_SCIENTIST: "DataScientist",
  NETWORK_ADMINISTRATOR: "NetworkAdministrator",
  POWER_USER_ACCESS: "PowerUserAccess",
  READON_ONLY_ACCESS: "ReadOnlyAccess",
  SECURITY_AUDIT: "SecurityAudit",
  SUPPORT_USER: "SupportUser",
  SYSTEMK_ADMINISTRATOR: "SystemAdministrator",
  VIEW_ONLY_ACCESS: "ViewOnlyAccess",
} as const;
