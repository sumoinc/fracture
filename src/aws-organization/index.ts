export * from "./aws-account";
export * from "./aws-environment";
export * from "./aws-organization";
export * from "./aws-organization";
export * from "./aws-profile";
export * from "./aws-region";
