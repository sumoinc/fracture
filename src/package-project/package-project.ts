import { Project } from "projen";
import { NodePackageManager } from "projen/lib/javascript";
import {
  TypeScriptProject,
  TypeScriptProjectOptions,
} from "projen/lib/typescript";
import { Jest } from "../tests/jest";

export interface PackageProjectOptions
  extends Omit<TypeScriptProjectOptions, "defaultReleaseBranch"> {
  // readonly parent: MonorepoProject;
  // readonly testingOptions?: TestingOptions;
}

/**
 * Package Project
 */
export class PackageProject extends TypeScriptProject {
  /**
   * Returns all packages in the monorepo.
   */
  public static all(project: Project): Array<PackageProject> {
    const isDefined = (p: Project): p is PackageProject =>
      p instanceof PackageProject;
    return project.root.subprojects.filter(isDefined);
  }

  /**
   * The testing settings for this project.
   */
  // public testing?: Testing;

  constructor(options: PackageProjectOptions) {
    /**
     * Is this a root level project? This determines what folder structure to
     * use when storing the project.
     */
    //const isRootProject: boolean = options.parent === undefined;

    const outdir: string = options.parent ? `packages/${options.name}` : "";

    const artifactsDirectory: string = options.parent ? "dist" : "dist";

    /**
     * Builds a cachable output path at the root of the monorepo for dist
     * and other deployable artifacts.
     */
    /*
    const buildOutputPath = (name: string) => {
      return join(
        "../..",
        options.parent.deploymentArtifactRoot,
        `packages/${options.name}`,
        name
      );
    };*/

    super({
      /*************************************************************************
       * INPUTS
       ************************************************************************/

      ...options,

      /*************************************************************************
       * FORCED OPTIONS
       ************************************************************************/

      /**
       * Use main branch. It's 2023 by the way.
       */
      defaultReleaseBranch: "main",

      /**
       * Enable prettier for some great formatting.
       */
      prettier: true,

      /**
       * PNPM 8 all the way baby!
       */
      packageManager: NodePackageManager.PNPM,
      pnpmVersion: "8",

      /**
       * No license for now.
       */
      licensed: false,

      /**
       * Turn off the native jest support, we will be adding our own Jest
       * config beflow.
       */
      jest: false,

      /**
       * All package projects are placed into the packages path structure.
       */
      // outdir: `packages/${options.name}`,
      outdir,

      /**
       * Always package things in packages folder
       */
      package: true,

      /**
       * Final distributable assets go to a common directory in the root.
       */
      // artifactsDirectory: buildOutputPath("dist"),
      artifactsDirectory,
    });

    /***************************************************************************
     *
     * PROJEN SETUP
     *
     **************************************************************************/

    [
      ".DS_Store",
      ".gitattributes",
      ".prettierignore",
      ".prettierrc.json",
    ].forEach((f) => {
      this.addPackageIgnore(f);
    });
    this.addGitIgnore(".DS_Store");

    /***************************************************************************
     *
     * SETUP TESTING
     *
     **************************************************************************/

    new Jest(this);

    /***************************************************************************
     *
     * CONFIGURE TURBO
     *
     **************************************************************************/

    // if (Turbo.isActive(this)) {
    //   const turbo = Turbo.of(this)!;

    //   /**
    //    * Testing uses ts-jest so it doesn't depend on anyother tasks.
    //    */
    //   turbo.addTask(this, {
    //     projenTask: this.testTask,
    //     cachable: true,
    //     inputPaths: ["jest.config.json", ...Turbo.commonInputs(this)],
    //     outputPaths: ["coverage/**", "test-reports/**"],
    //   });

    //   /**
    //    * Compile and Package need to happen in a particular order.
    //    */
    //   const turboCompile = turbo.addTask(this, {
    //     projenTask: this.compileTask,
    //     cachable: true,
    //     inputPaths: [...Turbo.commonInputs(this)],
    //     outputPaths: [`${this.libdir}/**`],
    //   });
    //   const turboPackage = turbo.addTask(this, {
    //     projenTask: this.packageTask,
    //     cachable: true,
    //     inputPaths: [...Turbo.commonInputs(this)],
    //     outputPaths: [`${this.artifactsDirectory}/**`],
    //   });
    //   turboCompile.thenRun(turboPackage);
    // }
  }
}
