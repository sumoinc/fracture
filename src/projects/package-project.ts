import { Project } from "projen";
import { NodePackageManager } from "projen/lib/javascript";
import {
  TypeScriptProject,
  TypeScriptProjectOptions,
} from "projen/lib/typescript";
import { commonProjectSetup } from "./common";
import { Jest } from "../tests/jest";

export interface PackageProjectOptions
  extends Omit<TypeScriptProjectOptions, "defaultReleaseBranch"> {
  // readonly parent: MonorepoProject;
  // readonly testingOptions?: TestingOptions;
}

/**
 * Package Project
 */
export class PackageProject extends TypeScriptProject {
  /**
   * Returns all packages in the monorepo.
   */
  public static all(project: Project): Array<PackageProject> {
    const isDefined = (p: Project): p is PackageProject =>
      p instanceof PackageProject;
    return project.root.subprojects.filter(isDefined);
  }

  /**
   * The testing settings for this project.
   */
  // public testing?: Testing;

  constructor(options: PackageProjectOptions) {
    /**
     * Is this a root level project? This determines what folder structure to
     * use when storing the project.
     */
    //const isRootProject: boolean = options.parent === undefined;

    const outdir: string = options.parent ? `packages/${options.name}` : "";

    const artifactsDirectory: string = options.parent ? "dist" : "dist";

    /**
     * Builds a cachable output path at the root of the monorepo for dist
     * and other deployable artifacts.
     */
    /*
    const buildOutputPath = (name: string) => {
      return join(
        "../..",
        options.parent.deploymentArtifactRoot,
        `packages/${options.name}`,
        name
      );
    };*/

    super({
      /*************************************************************************
       * DEFAULTS
       ************************************************************************/

      /**
       * No license by default.
       */
      licensed: false,

      /**
       * Use Node 18.x by default for package builds
       */
      workflowNodeVersion: "18",

      /**
       * Use main branch. It's 2023 by the way.
       */
      defaultReleaseBranch: "main",

      /**
       * Enable prettier for some great formatting.
       */
      prettier: true,

      /**
       * PNPM 8 all the way baby!
       */
      packageManager: NodePackageManager.PNPM,
      pnpmVersion: "8",

      /**
       * Turn off the native jest support, we will be adding our own Jest
       * config beflow.
       */
      jest: false,

      /**
       * All package projects are placed into the packages path structure.
       */
      // outdir: `packages/${options.name}`,
      outdir,

      /**
       * Always package assets for distribution.
       */
      package: true,

      /**
       * Final distributable assets go to a common directory in the root.
       */
      // artifactsDirectory: buildOutputPath("dist"),
      artifactsDirectory,

      /*************************************************************************
       * INPUTS
       ************************************************************************/

      ...options,
    });

    /***************************************************************************
     *
     * COMMON SETUP
     *
     **************************************************************************/

    commonProjectSetup(this);

    /***************************************************************************
     *
     * SETUP IGNORES
     *
     **************************************************************************/

    [
      ".DS_Store",
      ".gitattributes",
      ".prettierignore",
      ".prettierrc.json",
      ".projenrc.ts",
      "node_modules",
      "/**/*.spec.*",
    ].forEach((f) => {
      this.addPackageIgnore(f);
    });

    [".DS_Store"].forEach((f) => {
      this.addGitIgnore(f);
    });

    /***************************************************************************
     *
     * SETUP TESTING
     *
     **************************************************************************/

    new Jest(this);

    /***************************************************************************
     *
     * CONFIGURE TURBO
     *
     **************************************************************************/

    // if (Turbo.isActive(this)) {
    //   const turbo = Turbo.of(this)!;

    //   /**
    //    * Testing uses ts-jest so it doesn't depend on anyother tasks.
    //    */
    //   turbo.addTask(this, {
    //     projenTask: this.testTask,
    //     cachable: true,
    //     inputPaths: ["jest.config.json", ...Turbo.commonInputs(this)],
    //     outputPaths: ["coverage/**", "test-reports/**"],
    //   });

    //   /**
    //    * Compile and Package need to happen in a particular order.
    //    */
    //   const turboCompile = turbo.addTask(this, {
    //     projenTask: this.compileTask,
    //     cachable: true,
    //     inputPaths: [...Turbo.commonInputs(this)],
    //     outputPaths: [`${this.libdir}/**`],
    //   });
    //   const turboPackage = turbo.addTask(this, {
    //     projenTask: this.packageTask,
    //     cachable: true,
    //     inputPaths: [...Turbo.commonInputs(this)],
    //     outputPaths: [`${this.artifactsDirectory}/**`],
    //   });
    //   turboCompile.thenRun(turboPackage);
    // }
  }
}
