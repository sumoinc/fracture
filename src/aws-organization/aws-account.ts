import { Component } from "projen";
import { ValueOf } from "type-fest";
import { AwsEnvironment, IAwsEnvironment } from "./aws-environment";
import { AwsOrganization } from "./aws-organization";
import { AwsRegionIdentifier } from "./aws-region";
import { CdkProject } from "../cdk";
import { DeployTarget } from "../cdk/config/deploy-target";
import {
  GithubOidcRole,
  GithubOidcRoleOptions,
  IGithubOidcRole,
} from "../github/github-oidc-role";

export interface IAwsAccount {
  readonly orgId: string;
  readonly account: string;
  readonly environments: Array<IAwsEnvironment>;
  //readonly deployTargets: Array<IDeployTarget>;
  readonly githubOidcRoles: Array<IGithubOidcRole>;
}

export interface AwsAccountOptions {
  readonly orgId?: string;
  readonly account: string;
  readonly regions?: Array<ValueOf<typeof AwsRegionIdentifier>>;
  /**
   * Define the of releases allowed into this account.
   */
  readonly githubOidcRoleOptions?: Array<
    Omit<GithubOidcRoleOptions, "account">
  >;
}

export class AwsAccount extends Component implements IAwsAccount {
  public static all(project: CdkProject): Array<AwsAccount> {
    const isDefined = (c: Component): c is AwsAccount =>
      c instanceof AwsAccount;
    return project.components.filter(isDefined);
  }

  /**
   * Create account if needed or return it if not.
   */
  public static ensureExists(
    project: CdkProject,
    options: AwsAccountOptions
  ): AwsAccount {
    /**
     * Find function
     */
    const isDefined = (c: Component): c is AwsAccount =>
      c instanceof AwsAccount && c.account === options.account;

    /**
     * It should exist in the project supplied.
     * Return this instance
     */
    return (
      project.components.find(isDefined) ?? new AwsAccount(project, options)
    );
  }

  public readonly orgId: string;
  public readonly account: string;

  constructor(public readonly project: CdkProject, options: AwsAccountOptions) {
    super(project);

    /***************************************************************************
     *
     * DEFAULTS
     *
     **************************************************************************/

    this.orgId =
      options.orgId ??
      AwsOrganization.byAccount(project, options.account).orgId;
    this.account = options.account;

    options.regions?.forEach((region) => {
      AwsEnvironment.ensureExists(this.project, {
        orgId: this.orgId,
        account: this.account,
        region,
      });
    });
  }

  public get environments() {
    return AwsEnvironment.all(this.project).filter(
      (env) => env.account === this.account
    );
  }

  public get deployTargets() {
    return DeployTarget.all(this.project).filter(
      (env) => env.account === this.account
    );
  }

  public get githubOidcRoles() {
    return GithubOidcRole.all(this.project).filter(
      (env) => env.account === this.account
    );
  }

  public config(): IAwsAccount {
    return {
      account: this.account,
      orgId: this.orgId,
      environments: this.environments.map((e) => e.config()),
      //deployTargets: this.deployTargets.map((t) => t.config()),
      githubOidcRoles: this.githubOidcRoles.map((r) => r.config()),
    };
  }
}
