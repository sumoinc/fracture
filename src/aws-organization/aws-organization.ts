import { Component } from "projen";
import { ValueOf } from "type-fest";
import { AwsAccount } from "./aws-account";
import { AwsRegionIdentifier } from "./aws-region";
import { CdkProject } from "../projects";

export interface AwsOrganizationOptions {
  readonly orgId: string;
  readonly ssoStartUrl: string;
  readonly ssoAccount: string;

  /**
   * @default AwsRegionIdentifier.US_EAST_1
   */
  readonly ssoRegion?: ValueOf<typeof AwsRegionIdentifier>;
}

export class AwsOrganization extends Component {
  public static all(project: CdkProject): Array<AwsOrganization> {
    const isDefined = (c: Component): c is AwsOrganization =>
      c instanceof AwsOrganization;
    return project.components.filter(isDefined);
  }

  public readonly orgId: string;
  public readonly ssoStartUrl: string;
  public readonly ssoAccount: string;
  public readonly ssoRegion: ValueOf<typeof AwsRegionIdentifier>;

  constructor(
    public readonly project: CdkProject,
    options: AwsOrganizationOptions
  ) {
    super(project);

    /***************************************************************************
     *
     * DEFAULTS
     *
     **************************************************************************/

    this.orgId = options.orgId;
    this.ssoStartUrl = options.ssoStartUrl;
    this.ssoAccount = options.ssoAccount;
    this.ssoRegion = options.ssoRegion ?? AwsRegionIdentifier.US_EAST_1;
  }

  public get accounts() {
    return AwsAccount.all(this.project).filter((a) => a.orgId === this.orgId);
  }

  public hasAccount(account: string): boolean {
    return this.accounts.some((a) => a.account === account);
  }

}
