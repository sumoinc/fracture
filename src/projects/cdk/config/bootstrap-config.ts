import { Component } from "projen";
import { TypeScriptProject } from "projen/lib/typescript";
import { ValueOf } from "type-fest";
import { AwsRegionIdentifier } from "../../../aws-organization";

export const BootstrapFlag = {
  /**
   * Earlier versions of the bootstrap template created an AWS KMS key in each
   * bootstrapped environment by default. To avoid charges for the KMS key,
   * re-bootstrap these environments using --no-bootstrap-customer-key. The
   * current default is no KMS key, which helps avoid these charges.
   */
  NO_BOOTSTRAP_CUSTOMER_KEY: "--no-bootstrap-customer-key",

  /**
   * overrides the name of the Amazon S3 bucket. May require changes to your
   * CDK app (see Stack synthesizers).
   */
  BOOTSTRAP_BUCKET_NAME: "--bootstrap-bucket-name",

  /**
   * overrides the AWS KMS key used to encrypt the S3 bucket.
   */
  BOOTSTRAP_KMS_KEY_ID: "--bootstrap-kms-key-id",

  /**
   * specifies the ARNs of managed policies that should be attached to the
   * deployment role assumed by AWS CloudFormation during deployment of your
   * stacks. By default, stacks are deployed with full administrator
   * permissions using the AdministratorAccess policy.
   *
   * The policy ARNs must be passed as a single string argument, with the
   * individual ARNs separated by commas. For example:
   *
   * --cloudformation-execution-policies "arn:aws:iam::aws:policy/AWSLambda_FullAccess,arn:aws:iam::aws:policy/AWSCodeDeployFullAccess".
   *
   * To avoid deployment failures, be sure the policies that you specify are
   * sufficient for any deployments you will perform in the environment
   * being bootstrapped.
   */
  CLOUDFORMATION_EXECUTION_POLICIES: "--cloudformation-execution-policies",

  /**
   * a string that is added to the names of all resources in the bootstrap
   * stack. A qualifier lets you avoid resource name clashes when you provision
   * multiple bootstrap stacks in the same environment. The default is
   * hnb659fds (this value has no significance).
   *
   * Changing the qualifier also requires that your CDK app pass the changed
   * value to the stack synthesizer. For more information, see Stack
   * synthesizers.
   */
  QUALIFIER: "--qualifier",

  /**
   * Use the indicated AWS profile when bootstrapping
   *
   * [string]
   */
  PROFILE: "--profile",

  /**
   * adds one or more AWS CloudFormation tags to the bootstrap stack.
   */
  TAGS: "--tags",

  /**
   * lists the AWS accounts that may deploy into the environment being
   * bootstrapped.
   *
   * Use this flag when bootstrapping an environment that a CDK Pipeline in
   * another environment will deploy into. The account doing the bootstrapping
   * is always trusted.
   */
  TRUST: "--trust",

  /**
   * lists the AWS accounts that may look up context information from the
   * environment being bootstrapped.
   *
   * Use this flag to give accounts permission to synthesize stacks that will
   * be deployed into the environment, without actually giving them permission
   * to deploy those stacks directly.
   */
  TRUST_FOR_LOOKUP: "--trust-for-lookup",

  /**
   * Prevents the bootstrap stack from being deleted. For more information, see
   * Protecting a stack from being deleted in the AWS CloudFormation User Guide.
   */
  TERMINATION_PROTECTION: "--termination-protection",
} as const;

export const AwsEnvironmentBootstrapQualifier = {
  DEFAULT: "hnb659fds",
} as const;

export interface BootstrapFlags {
  [BootstrapFlag.NO_BOOTSTRAP_CUSTOMER_KEY]?: boolean;
  [BootstrapFlag.BOOTSTRAP_BUCKET_NAME]?: string;
  [BootstrapFlag.BOOTSTRAP_KMS_KEY_ID]?: string;
  [BootstrapFlag.CLOUDFORMATION_EXECUTION_POLICIES]?: string;
  [BootstrapFlag.QUALIFIER]:
    | ValueOf<typeof AwsEnvironmentBootstrapQualifier>
    | string;
  [BootstrapFlag.PROFILE]: string;
  [BootstrapFlag.TAGS]?: string;
  [BootstrapFlag.TRUST]?: string;
  [BootstrapFlag.TRUST_FOR_LOOKUP]?: string;
  [BootstrapFlag.TERMINATION_PROTECTION]?: boolean;
}

export type BootstrapConfigOptions = {
  readonly account: string;
  readonly region: ValueOf<typeof AwsRegionIdentifier>;

  /**
   * Earlier versions of the bootstrap template created an AWS KMS key in each
   * bootstrapped environment by default. To avoid charges for the KMS key,
   * re-bootstrap these environments using --no-bootstrap-customer-key. The
   * current default is no KMS key, which helps avoid these charges.
   *
   * Set to "true" to avoid charges for the KMS key.
   *
   * @default true
   */
  readonly noBootstrapCustomerKey?: boolean;

  /**
   * The qualifier for this CDK bootstraping instance. Uses the CDK default if
   * not supplied as arg.
   *
   * @default AwsEnvironmentBootstrapQualifier.DEFAULT
   */
  readonly qualifier?:
    | ValueOf<typeof AwsEnvironmentBootstrapQualifier>
    | string;

  /**
   * Local profile name to use when depploying the bootstrap config.
   */
  readonly profileName: string;

  /**
   * What version to use when bootstrapping?
   *
   * @default latest
   */
  readonly cdkVersion?: string;
};

export class BootstrapConfig extends Component {
  public readonly account: string;
  public readonly region: ValueOf<typeof AwsRegionIdentifier>;
  public readonly noBootstrapCustomerKey: boolean;
  public readonly qualifier:
    | ValueOf<typeof AwsEnvironmentBootstrapQualifier>
    | string;
  public readonly profileName: string;
  public readonly flags: BootstrapFlags;
  public readonly cdkVersion: string;

  constructor(
    public readonly project: TypeScriptProject,
    options: BootstrapConfigOptions
  ) {
    super(project);

    /***************************************************************************
     * DEFAULT OPTIONS
     **************************************************************************/

    this.account = options.account;
    this.region = options.region;
    this.noBootstrapCustomerKey = options.noBootstrapCustomerKey ?? true;
    this.qualifier =
      options.qualifier ?? AwsEnvironmentBootstrapQualifier.DEFAULT;
    this.profileName = options.profileName;
    this.cdkVersion = options.cdkVersion ?? "latest";

    this.flags = {
      [BootstrapFlag.NO_BOOTSTRAP_CUSTOMER_KEY]: this.noBootstrapCustomerKey,
      [BootstrapFlag.PROFILE]: this.profileName,
      [BootstrapFlag.QUALIFIER]: this.qualifier,
    };
  }

  public get bootstrapCommand() {
    const cdkCommand =
      this.cdkVersion === "latest" ? "cdk" : `cdk@${this.cdkVersion}`;
    return [
      "npx",
      cdkCommand,
      "bootstrap",
      `aws://${this.account}/${this.region}`,

      // add flags
      ...Object.entries(this.flags).map(([key, value]) => {
        return `${key}=${value}`;
      }),
    ].join(" ");
  }
}
