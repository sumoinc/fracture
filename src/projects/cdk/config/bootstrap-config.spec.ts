import { BootstrapConfig } from "./bootstrap-config";
import { AwsRegionIdentifier } from "../../../aws-organization";
import { PackageProject } from "../../package-project";

describe("Success conditions", () => {
  test("Smoke Test", () => {
    const project = new PackageProject({
      name: "my-project",
    });
    expect(project).toBeTruthy();

    // bootstrap an environment
    const config = new BootstrapConfig(project, {
      account: "123456789012",
      region: AwsRegionIdentifier.US_EAST_1,
      profileName: "foo",
    });

    expect(config).toBeTruthy();
  });
});

describe("Command output", () => {
  test("With CDK Version", () => {
    const project = new PackageProject({
      name: "my-project",
    });
    expect(project).toBeTruthy();

    // bootstrap an environment
    const config = new BootstrapConfig(project, {
      account: "123456789012",
      region: AwsRegionIdentifier.US_EAST_1,
      profileName: "foo",
      cdkVersion: "2.210.0",
    });

    expect(config.bootstrapCommand).toBeTruthy();
    expect(config.bootstrapCommand).toMatchSnapshot();

    // console.log(config.bootstrapCommand);
  });

  test("Without CDK Version", () => {
    const project = new PackageProject({
      name: "my-project",
    });
    expect(project).toBeTruthy();

    // bootstrap an environment
    const config = new BootstrapConfig(project, {
      account: "123456789012",
      region: AwsRegionIdentifier.US_EAST_1,
      profileName: "foo",
    });

    expect(config.bootstrapCommand).toBeTruthy();
    expect(config.bootstrapCommand).toMatchSnapshot();

    // console.log(config.bootstrapCommand);
  });
});
