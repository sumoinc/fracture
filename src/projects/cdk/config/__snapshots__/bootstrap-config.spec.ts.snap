// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Command output With CDK Version 1`] = `"npx cdk@2.210.0 bootstrap aws://123456789012/us-east-1 --no-bootstrap-customer-key=true --profile=foo --qualifier=hnb659fds"`;

exports[`Command output Without CDK Version 1`] = `"npx cdk@latest bootstrap aws://123456789012/us-east-1 --no-bootstrap-customer-key=true --profile=foo --qualifier=hnb659fds"`;
